### PR TITLE
Fix #89: Invoke a hook before/after generating a page

### LINKED ISSUE #89: Invoke a hook before/after generating a page
[Issuehunt badges](https://issuehunt.io/r/egoist/saber/issues/89)

#### What problem does this feature solve?

Would be useful for https://github.com/egoist/saber/issues/71

#### What does the proposed API look like?

```js
api.hooks.afterGeneratePage.tapPromise(ID, context => {
  const amp = shouldGenerateAMPPage(context)
  if (amp) {
    //...
  }
})
```

#### How should this be implemented in your opinion?

N/A

#### Are you willing to work on this yourself?

PR welcome.

> There is a $20.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/egoist/saber/issues/89).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/egoist/saber/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by MarchWorks**
**Summary**

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI/CSS related code, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: packages/saber/vue-renderer/lib/index.js [hunk 1]
```diff
@@ -276,7 +276,9 @@ class VueRenderer {
               .replace(/^\s+/gm, '')
               .replace(/\n+</g, '<')
               .replace('<div id="_saber"></div>', markup)
+            await this.api.hooks.beforeGeneratePage.promise(context)
             await fs.outputFile(route.outputFilePath, html, 'utf8')
+            await this.api.hooks.afterGeneratePage.promise(context)
```
Comment (egoist) [inline packages/saber/vue-renderer/lib/index.js]: let's also pass `{ html, outputFilePath }` to the hook as the second argument.

Comment (MarchWorks) [reply packages/saber/vue-renderer/lib/index.js]: If I am not mistaken I should name it here `afterGeneratePage: new AsyncSeriesHook(['context', 'name'])` what name should I use?

Comment (MarchWorks) [reply packages/saber/vue-renderer/lib/index.js]: maybe `file`

Comment (egoist) [reply packages/saber/vue-renderer/lib/index.js]: LGTM

**Comment by MarchWorks** [comment]
Don't merge ` packages/saber/lib/plugins/config-css.js` got deleted somehow

**Comment by MarchWorks** [comment]
@egoist LGTM

**Comment by MarchWorks** [comment]
Nice

**Pull request merged**